core/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -181,14 +189,54 @@ public int hashCode() {
         return Objects.hash(this.field, this.id, this.max);
     }
 
-    public Query toFilter(QueryShardContext context) {
+    public Query toFilter(QueryShardContext context, int shardId, int numShards) {
         final MappedFieldType type = context.fieldMapper(field);
         if (type == null) {
             throw new IllegalArgumentException("field " + field + " not found");
         }
+
         if (UidFieldMapper.NAME.equals(field)) {
-            return new TermsSliceQuery(field, id, max);
+            if (numShards == 1) {
+                return new TermsSliceQuery(field, id, max);
+            }
+
+            if (max >= numShards) {
+                // the number of slices is greater than the number of shards
+                // in such case we can reduce the number of requested shards by slice
+
+                // first we check if the slice is responsible of this shard
+                int targetShard = id % numShards;
```

Comment:
If id or numShards are negative we're screwed anyway so using floorMod would just hide the problem ? We always check that id is positive so it should not be a problem.